Retry alternative

rpoelstra opened the pull request:
Hi,

I've looked at your retry scheme. I would like to propose an alternative. With your implementation I get messages about the loop being delayed too long. My implementation is less aggressive as it employs a larger delay between writes (I intend to make the delay configurable, just as poll_interval, so it can be set close to 0 in the future). It also ignores failed STATUS commands, as they normally aren't essential to be executed properly each time. Other commands are retried indefinitely until a success or a different command is issued. This will make it less likely that important commands such as 'CLOSE' are lost.

This PR is based on a slightly older master, but I can update it to be based on the latest commit if we go forward with this.

What do you think about this?

rrooggiieerr commented:
I like the idea of using the poll interval for retrying. Of course if a poll interval of 0 (continuous polling) is used the retry interval could also be 0.

Now I'm thinking about it the `OPEN`, `CLOSE` and `STOP` commands make the window opener **do** something while the `STATUS`, `DEVICE` and `VERSION` commands request something. Where for `DEVICE` and `VERSION` the returned string is used and for `STATUS` the returned integer.

`OPEN`, `CLOSE` and `STOP` should really be successful and otherwise retried. `DEVICE` and `VERSION` are only used on setup of the ESP and it would be nice if they are successful but not crucial. Retrying those commands don't do much harm, it would only slow down the setup process a bit. A successful execution of `STATUS` is only required at setup since it validates the stored state and is used to detect a power outage. If the `STATUS` and stored state are not equal this means power outage has happened. Again retrying during setup would not do much harm.

Then we got the `STATUS` command during normal operation for which it is not really crucial if it's successful or not.
 
I think you could just put the to be executed `OPEN`, `CLOSE` or `STOP` command in a global variable end if that variable is not `nullptr` retry the command. on successful execution you put `nullptr` in the variable.

```
if(this->cmd_ == nullptr) {
	//Update @ poll_interval
	poll_interval = this->polling_interval_;
} else {
	//Update @ retry_interval
	poll_interval = 1000;
}
```

And then later
```
if(this->cmd_ == nullptr) {
	response = this->send_cmd_(AXACommand::STATUS);
} else {
	response = this->send_cmd_(this->cmd_);
	if (response >= int(AXAResponseCode::OK) && response <= int(AXAResponseCode::Error))
		this->cmd_ = nullptr;
}
```

Or something in that direction.

By the way, what are the `DEVICE` and `VERSION` of your AXA Remote 2.0S? Can it be used to distinct te device from the non synchronised version?

rpoelstra commented:
I appreciated that you like my line of thinking.

I've updated the code to match the latest master and made the retry interval a configuration parameter. I didn't yet update the setup() function to retry commands where needed.

I've added the `try_cmd()` function to express at the call site whether a command is important and should be retried or whether a single shot would be sufficient. Also without this additional function the instance variable to store the current command would need to be set at each call, which isn't DRY.

I'm happy to look at retrying in the setup code, but there is a lot of commented-out code which feels kind of necessary? You also set the `power_outage_detected_` var, but never clear it. Is that intentional?

The versions as reported by my device:
[10:08:12][C][axaremote.cover:064]:   Device: AXA RV2900 2.0
[10:08:12][C][axaremote.cover:065]:   Firmware: Firmware V2.00 2013

I'm not sure what a single motor would report?

rpoelstra commented:
I was thinking, maybe we should swap the names of `send_cmd` and 'try_send_cmd`, as that would better reflect their purpose. If you agree, I'll sharpen my find-and-replace-fu.

rrooggiieerr commented:
Following up on our discussion I also created a retry alternative that I put in the [retry_alternative branche ](https://github.com/rrooggiieerr/esphome-axaremote/tree/retry_alternative)

Let me know your thoughts

rrooggiieerr commented:
I think this can be explained

The AXA did receive the OPEN command and acts on it, but the expected response has interfered with a synchronisation message and the code thinks the command failed.

5 seconds later the STATUS command returns the unlocked state but the code was still expecting the AXA to be closed (because it thinks the OPEN command failed). This makes the code think the AXA has been opened by the remote.

What I don't understand is why the retrying is not done in 1 second intervals, assuming you use a polling interval greater than 1 second

rrooggiieerr commented:
I think I've fixed the retrying

rrooggiieerr commented:
That's good to hear!

Please give it a bit more testing, and if you see room for improvements let me know.

I don't think there is a need to make the retry interval configurable. It's now 1 second or less if the polling interval is less.

On the normal AXA Remote 2.0 with no polling interval configured this means retries are done straight away, and on your AXA Remote 2.0S with a 5 second polling interval configured this means retries are done every second.

The retries, and polling, are now also done related to the last command send, and not the last retry or poll. Thus in your case a next poll is 5 seconds after the last command is send. This is in most cases the STATUS command, but if you start a direction then it's 5 seconds after the STOP, OPEN or CLOSE command. I hope this makes sense.

Maybe you could also test if shorter polling intervals are possible with the AXA Remote 2.0S. A shorter polling interval will make the code a bit more responsive.

Something else that would be interesting is if we could use the synchronisation messages by the AXA Remote to retrieve the position. I've just added some additional logging that shows the pre-command serial buffer contents.

rpoelstra commented:
Until now it's been working fine....
For your information: I've been using an ESP32 for the development, but now all three systems use an ESP8266 and that also works fine.
I think we can close this PR and merge your branch?
